model/godleyTableWindow.cc

```diff
@@ -825,8 +825,10 @@ namespace minsky
   void GodleyTableWindow::pushHistory()
   {
     while (history.size()>maxHistory) history.pop_front();
-    if (history.empty() || history.back()!=godleyIcon->table.getData())
-      history.push_back(godleyIcon->table.getData());
+    // Include asset class of each column in history at avoid spurious noAssetClass columns. For ticket 1118.
+    if (history.empty() || (history.back().first!=godleyIcon->table.getData() && history.back().second!=godleyIcon->table._assetClass())) {
+      history.push_back(make_pair(godleyIcon->table.getData(), godleyIcon->table._assetClass()));
```

Comment:
This is a fair point, however I believe it would be better to push the entire table object onto history. Not only does it have these two items (data, asset classes) but also title and the deprecated doubleEntryCompliant flag (which we've still to get rid of).

Reply:
That's a very good suggestion, thanks. I will implement this today. The test t0001a.sh fails for some weird reason, which doesn't make sense, because all I've done is to stash the asset class of each column. In my development folder, the test is failing all over the place suddenly. I don't know what changes I made to cause it. I reverted the last few commits to see, but it didn't make a difference...

Reply:
Implementing the above idea shortens the code dramatically, except that it does not do anything. Because it is very short I put it here:

godleyTableWindow.h:

`std::deque<GodleyTable> history;`

godleyTableWindow.cc:

`  void GodleyTableWindow::pushHistory()
  {
    while (history.size()>maxHistory) history.pop_front();
    if (history.empty() || (&history.back()!=&godleyIcon->table)) {
      history.push_back(godleyIcon->table);
    }
    historyPtr=history.size();
  }
      
  void GodleyTableWindow::undo(int changes)
  {
    if (historyPtr==history.size())
      pushHistory();
    historyPtr-=changes;
    if (historyPtr > 0 && historyPtr <= history.size())
      {
        auto& d=history[historyPtr-1];
        if (!&d) return; // should not happen
	godleyIcon->table=d;
        requestRedraw();
      }
  }`

Then, I have also tried to pass a "currTable" in pushHistory(), which also required editing minskyTCLObj.h on line 53:

`        if (!t || (!t->is_const && (!t->is_setterGetter || argc>1)))
            if (auto gtw=t->memberPtrCasted<GodleyTableWindow>()) {
              auto g=t->memberPtrCasted<GodleyTable>(); 
              gtw->pushHistory(*g);
            }`

Then, in godleyTableWindow.h:

`void pushHistory(const GodleyTable& currTable);`

and in godleyTableWindow.cc:

`  void GodleyTableWindow::pushHistory(const GodleyTable& currTable)
  {
    while (history.size()>maxHistory) history.pop_front();
    if (history.empty() || (&history.back()!=&currTable)) {
      history.push_back(currTable);
    }
    historyPtr=history.size();
  }
      
  void GodleyTableWindow::undo(int changes)
  {
    if (historyPtr==history.size())
      pushHistory(godleyIcon->table);
    historyPtr-=changes;
    if (historyPtr > 0 && historyPtr <= history.size())
      {
        auto& d=history[historyPtr-1];
        if (!&d) return; // should not happen
	godleyIcon->table=d;
        requestRedraw();
      }
  }`

This produces a segmentation fault that I think has to do with the fact that the objects in gtw->pushHistory(*g) in minskyTCLObj.h and pushHistory(godleyIcon->table) in godleyWindowTable::undo() are not the same? From backtrace in gdb:

(gdb) frame 7
#7  0x00005555556a962a in minsky::member_entry_hook<char const**> (argc=2, argv=0x55555656be90) at gui-tk/minskyTCLObj.h:53
**53	              gtw->pushHistory(*g);**
(gdb) frame 6
#6  0x0000555555b3d9a4 in minsky::GodleyTableWindow::pushHistory (this=0x555556b75400, currTable=...) at model/godleyTableWindow.cc:831
**831	      history.push_back(currTable);**

Reply:
I've tried several approaches, including using a weak_ptr for the history, and using either the GodleyTable or GodleyIcon objects to define the history, but when it successfully compiles, nothing happens when I press ctrl+Z, the changes cannot be undone. I think my lack of understanding of how these objects work prevents me from replacing the actual godleyIcon or godleyIcon->table objects with one of the elements of the history object. 

On a positive note, I've managed to fix my previous attempt (using a std::deque of GodleyTable::Data and vector<GodleyAssetClass::AssetClass> pairs) so that it now passes all the tests. I hope it fixes at least some of the Godley table bugs.

Reply:
> Implementing the above idea shortens the code dramatically, except that it does not do anything. Because it is very short I put it here:
> 
> godleyTableWindow.h:
> 
> `std::deque<GodleyTable> history;`
> 
> godleyTableWindow.cc:
> 
> ` void GodleyTableWindow::pushHistory()
> {
> while (history.size()>maxHistory) history.pop_front();
> if (history.empty() || (&history.back()!=&godleyIcon->table)) {

It seems like mentally you think you are storing pointers to the Godley table. Whereas. you are actually making deep copies. Which you have to do - storing pointers won't work. At this point you need to test for deep equality, not identity. Now the compiler doesn't automatically generate an equality operator, (thats coming in C++-20) so you will need to write one yourself. Fortunately, its pretty easy. 

The net effect now is that history is probably full of the same state, as the top of the history stack will always be a different object to the current table, even though they may be in the same state.

> history.push_back(godleyIcon->table);
> }
> historyPtr=history.size();
> }
> 
> void GodleyTableWindow::undo(int changes)
> {
> if (historyPtr==history.size())
> pushHistory();
> historyPtr-=changes;
> if (historyPtr > 0 && historyPtr <= history.size())
> {
> auto& d=history[historyPtr-1];
> if (!&d) return; // should not happen

Again - because we're not storing pointers, this statement makes no sense whatsoever.

> godleyIcon->table=d;
> requestRedraw();
> }
> }`
> 
> Then, I have also tried to pass a "currTable" in pushHistory(), which also required editing minskyTCLObj.h on line 53:
> 
> ` if (!t || (!t->is_const && (!t->is_setterGetter || argc>1))) if (auto gtw=t->memberPtrCasted<GodleyTableWindow>()) { auto g=t->memberPtrCasted<GodleyTable>(); gtw->pushHistory(*g); }`
> 
> Then, in godleyTableWindow.h:
> 
> `void pushHistory(const GodleyTable& currTable);`
> 
> and in godleyTableWindow.cc:
> 
> ` void GodleyTableWindow::pushHistory(const GodleyTable& currTable)
> {
> while (history.size()>maxHistory) history.pop_front();
> if (history.empty() || (&history.back()!=&currTable)) {
> history.push_back(currTable);
> }
> historyPtr=history.size();
> }
> 
> void GodleyTableWindow::undo(int changes)
> {
> if (historyPtr==history.size())
> pushHistory(godleyIcon->table);
> historyPtr-=changes;
> if (historyPtr > 0 && historyPtr <= history.size())
> {
> auto& d=history[historyPtr-1];
> if (!&d) return; // should not happen
> godleyIcon->table=d;
> requestRedraw();
> }
> }`
> 
> This produces a segmentation fault that I think has to do with the fact that the objects in gtw->pushHistory(*g) in minskyTCLObj.h and pushHistory(godleyIcon->table) in godleyWindowTable::undo() are not the same? From backtrace in gdb:
> 
> (gdb) frame 7
> #7 0x00005555556a962a in minsky::member_entry_hook<char const**> (argc=2, argv=0x55555656be90) at gui-tk/minskyTCLObj.h:53
> *_53 gtw->pushHistory(_g);__
> (gdb) frame 6
> #6 0x0000555555b3d9a4 in minsky::GodleyTableWindow::pushHistory (this=0x555556b75400, currTable=...) at model/godleyTableWindow.cc:831
> **831 history.push_back(currTable);**

Reply:
Thank you, I really didn't understand why it wasn't working. 

I will add a bool operator==(const GodleyTable& x) const; in godleyTableWindow, test it and tomorrow morning commit the changes.